src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKeySynchronizer.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class AccessKeySynchronizer : IAccessKeySynchronizer, IDisposable
+    {
+        private readonly ConcurrentDictionary<ServiceEndpoint, object> _endpoints = new ConcurrentDictionary<ServiceEndpoint, object>(new Comparer());
```

Comment:
The service endpoints got by `AccessKeySynchronizer` should be unique instances, you can just use a list to store them. And you can replace the whole list during each update.